src/main/java/fi/aalto/cs/apluscourses/intellij/actions/APlusAuthenticationAction.java

```diff
@@ -2,21 +2,45 @@
 
 import com.intellij.openapi.actionSystem.AnActionEvent;
 import com.intellij.openapi.project.DumbAwareAction;
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.dal.APlusExerciseDataSource;
+import fi.aalto.cs.apluscourses.intellij.services.Dialogs;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
 import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
-import fi.aalto.cs.apluscourses.presentation.APlusAuthenticationViewModel;
+import fi.aalto.cs.apluscourses.presentation.AuthenticationViewModel;
 import fi.aalto.cs.apluscourses.presentation.MainViewModel;
-import fi.aalto.cs.apluscourses.ui.APlusAuthenticationView;
 import org.jetbrains.annotations.NotNull;
 
 public class APlusAuthenticationAction extends DumbAwareAction {
+
+  private final MainViewModelProvider mainViewModelProvider;
+
+  private final Dialogs dialogs;
+
+  public APlusAuthenticationAction() {
+    this(PluginSettings.getInstance(), Dialogs.DEFAULT);
+  }
+
+  public APlusAuthenticationAction(MainViewModelProvider mainViewModelProvider, Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.dialogs = dialogs;
+  }
+
   @Override
   public void actionPerformed(@NotNull AnActionEvent e) {
-    MainViewModel mainViewModel = PluginSettings.getInstance().getMainViewModel(e.getProject());
-    APlusAuthenticationViewModel authenticationViewModel
-        = new APlusAuthenticationViewModel(e.getProject());
-    new APlusAuthenticationView(authenticationViewModel).show();
-    if (authenticationViewModel.getAuthentication() != null) {
-      mainViewModel.authenticationViewModel.set(authenticationViewModel);
+    Project project = e.getProject();
+    MainViewModel mainViewModel = mainViewModelProvider.getMainViewModel(project);
+    AuthenticationViewModel authentication = new AuthenticationViewModel();
+    if (dialogs.create(authentication, project).showAndGet()) {
+      mainViewModel.setExerciseDataSource(APlusExerciseDataSource::new, authentication::build);
     }
   }
+
+  public MainViewModelProvider getMainViewModelProvider() {
+    return mainViewModelProvider;
+  }
+
+  public Dialogs getDialogs() {
+    return dialogs;
+  }
```

Comment:
This is a very minor (and @superseacat would most likely strongly disagree with me here :laughing:), but I wouldn't add getters just to test the default constructor, instead I would just test how the `actionPerformed` of a default-constructed `APlusAuthenticationAction` behaves

Reply:
Actually never mind, since the default `actionPerformed` probably can't be tested since it uses all kinds of platform stuff